File: src/test/java/org/hisp/dhis/rules/functions/RuleFunctionFloorTests.java
package org.hisp.dhis.rules.functions;

import org.hamcrest.CoreMatchers;
import org.hamcrest.MatcherAssert;
import org.hisp.dhis.parser.expression.antlr.ExpressionParser;
import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.junit.Before;
import org.hisp.dhis.parser.expression.antlr.ExpressionParser;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

 import org.hamcrest.CoreMatchers;
import static org.mockito.Mockito.when;



@RunWith( MockitoJUnitRunner.class )
public class RuleFunctionFloorTests
{

        @Mock
        private ExpressionParser.ExprContext context;

        @Mock
        private CommonExpressionVisitor visitor;

        @Mock
        private ExpressionParser.ExprContext mockedFirstExpr;

        @Before
        public void setUp() {
                when(context.expr(0)).thenReturn( mockedFirstExpr );
        }

        @Test
        public void evaluateMustReturnFlooredValue()
        {
                RuleFunctionFloor floor = new RuleFunctionFloor();

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "4.1" );
                MatcherAssert.assertThat(floor.evaluate(context, visitor), CoreMatchers.<Object>is("4"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "0.8" );
                MatcherAssert.assertThat(floor.evaluate(context, visitor), CoreMatchers.<Object>is("0"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "5.1" );
                MatcherAssert.assertThat(floor.evaluate(context, visitor), CoreMatchers.<Object>is("5"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "1.0" );
                MatcherAssert.assertThat(floor.evaluate(context, visitor), CoreMatchers.<Object>is("1"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "-9.3" );
                MatcherAssert.assertThat(floor.evaluate(context, visitor), CoreMatchers.<Object>is("-10"));

                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "-5.9" );
                MatcherAssert.assertThat(floor.evaluate(context, visitor), CoreMatchers.<Object>is("-6"));
        }

        @Test
        public void return_zero_when_number_is_invalid()
        {
                when( visitor.castStringVisit( mockedFirstExpr ) ).thenReturn( "not a number" );

                RuleFunctionFloor floor = new RuleFunctionFloor();

                MatcherAssert.assertThat( floor.evaluate( context, visitor ), CoreMatchers.<Object>is("0") );
        }
}

File: src/test/java/org/hisp/dhis/rules/functions/RuleFunctionFloorTests.kt
package org.hisp.dhis.rules.functions

import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.hisp.dhis.rules.RuleVariableValue
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.util.*
import kotlin.test.assertFailsWith

@RunWith(JUnit4::class)
class RuleFunctionFloorTests {

    private val variableValues = HashMap<String, RuleVariableValue>()
    private val floorFunction = RuleFunctionFloor.create()

    @Test
    fun return_argument_rounded_down_to_nearest_whole_number() {

        assertThat(floorFunction.evaluate(listOf("0"), variableValues, null), `is`("0"))
        assertThat(floorFunction.evaluate(listOf("0.8"), variableValues, null), `is`("0"))
        assertThat(floorFunction.evaluate(listOf("1.0"), variableValues, null), `is`("1"))
        assertThat(floorFunction.evaluate(listOf("-9.3"), variableValues, null), `is`("-10"))
        assertThat(floorFunction.evaluate(listOf("5.9"), variableValues, null), `is`("5"))
        assertThat(floorFunction.evaluate(listOf("5"), variableValues, null), `is`("5"))
        assertThat(floorFunction.evaluate(listOf("-5"), variableValues, null), `is`("-5"))
    }

    @Test
    fun throw_illegal_argument_exception_when_argument_count_is_greater_than_expected() {
        assertFailsWith<IllegalArgumentException> {
            RuleFunctionFloor.create().evaluate(listOf("5.9", "6.8"), variableValues, null)
        }
    }

    @Test
    fun throw_illegal_argument_exception_when_arguments_count_is_lower_than_expected() {
        assertFailsWith<IllegalArgumentException> {
            RuleFunctionFloor.create().evaluate(ArrayList(), variableValues, null)
        }
    }

    @Test
    fun throw_null_pointer_exception_when_arguments_is_null() {
        assertFailsWith<NullPointerException> {
            RuleFunctionFloor.create().evaluate(null!!, variableValues, null)
        }
    }
}
